Add command-line arguments and hierarchical configuration files

connorjayr opened the pull request:
Adds the `clap` crate, which makes it easy to parse command-line arguments that are passed to the server binary. Also adds the `config` crate, which we use to read non-secret configuration files from a hierarchy of files and environment variables. The precedence of configuration sources is listed in `server/src/config.rs`.

connorjayr committed:
Add command-line arguments and hierarchical configuration files

Adds the `clap` crate, which makes it easy to parse command-line arguments that are passed to the server binary. Also adds the `config` crate, which we use to read non-secret configuration files from a hierarchy of files and environment variables. The precedence of configuration sources is listed in `server/src/config.rs`.

jputlock commented on server/src/config.rs:
```diff
@@ -0,0 +1,33 @@
+    pub fn new() -> Result<Self, ConfigError> {
+        let rust_env = env::var("RUST_ENV").unwrap_or_else(|_| "production".to_string());
+        config::Config::builder()
+            .add_source(File::with_name("config/default"))
```
do we want to replace `add_source` with `set_default` here for the default configuration?

connorjayr replied:
It looks like [`Config::set_default`](https://docs.rs/config/latest/config/builder/struct.ConfigBuilder.html#method.set_default) is meant for setting default configuration values individually rather than setting a default source, so I think it's probably best here to keep the defaults in a file. That's also what the crate seems to "recommend" in [their provided example](https://github.com/mehcode/config-rs/blob/master/examples/hierarchical-env/settings.rs#L50-L51).

connorjayr committed:
Don't set RUST_LOG in npm start

Log verbosity is set through command-line arguments, and defaults to INFO. So, we don't need to set it via an environment variable.

jputlock commented on server/src/main.rs:
```diff
@@ -1,53 +1,69 @@
+
+    let config = Config::new().expect("cannot read configuration");
+
+    let game_server = WsServer::new().start();
```
Why swap back to `new()` from `default()`?

connorjayr replied:
Just based on what I've seen in popular crates/other Rust code. I think people tend to use `new` when manually making a new instance of a struct, whereas `default` is used when trait constraints require a type that is `Default`, and are unaware of a conventional-but-not-guaranteed `new` method.

jputlock reviewed (approved):
Looks good, left a few comments.
